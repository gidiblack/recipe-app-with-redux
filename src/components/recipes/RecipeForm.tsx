import React from "react";
import { RecipeFormProps } from "../../types";
import SelectInput from "../common/SelectInput";
import TextInput from "../common/TextInput";

// this functional component renders the form used to manage recipes
const RecipeForm: React.FC<RecipeFormProps> = (props) => {
  const { recipe, authors, onChange, onSave, saving, errors } = props;
  return (
    <form onSubmit={onSave}>
      <h2>{recipe.id ? "Edit" : "Add"} Recipe</h2>
      {errors.onSave && (
        <div className="alert alert-danger" role="alert">
          {errors.onSave}
        </div>
      )}
      <TextInput
        name="title"
        label="Title"
        placeholder="Enter Recipe Title"
        onChange={onChange}
        value={recipe.title}
        error={errors.title}
      />

      <SelectInput
        name="authorId"
        label="Author"
        onChange={onChange}
        error={errors.authorId}
        defaultOption="Select Author"
        options={authors.map((author) => ({
          value: author.id,
          text: author.name,
        }))}
        value={recipe.authorId || " "}
      />

      <TextInput
        name="category"
        label="Category"
        placeholder="Enter Category"
        onChange={onChange}
        value={recipe.category}
        error={errors.category}
      />

      <button type="submit" disabled={saving} className="btn btn-primary">
        {saving ? "Saving..." : "Save"}
      </button>
    </form>
  );
};

export default RecipeForm;
