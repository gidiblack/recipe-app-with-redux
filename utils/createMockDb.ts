/* eslint-disable no-console */
const fs = require("fs");
const path2 = require("path");
const mockData = require("./mockData");

const { courses, authors } = mockData;
const data = JSON.stringify({ courses, authors });
const filepath = path2.join(__dirname, "db.json");

fs.writeFile(filepath, data, function (err) {
  err ? console.log(err) : console.log("Mock DB created.");
});
