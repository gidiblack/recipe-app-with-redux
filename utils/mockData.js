const recipeData = [
  {
    id: 1,
    title: "Grilled fish sticks",
    slug: "grilled-fish-sticks",
    authorId: 1,
    category: "Fish",
  },
  {
    id: 2,
    title: "Egusi Soup",
    slug: "egusi-soup",
    authorId: 3,
    category: "Soup",
  },
  {
    id: 3,
    title: "Jollof Rice",
    slug: "jollof-rice",
    authorId: 2,
    category: "Rice",
  },
  {
    id: 4,
    title: "Coconut Rice",
    slug: "coconut-rice",
    authorId: 2,
    category: "Rice",
  },
  {
    id: 5,
    title: "Fish Stew",
    slug: "fish-stew",
    authorId: 3,
    category: "Fish",
  },
  {
    id: 6,
    title: "Fried Chicken",
    slug: "fried-chicken",
    authorId: 1,
    category: "Chicken",
  },
  {
    id: 7,
    title: "Grilled Chicken",
    slug: "grilled-chicken",
    authorId: 2,
    category: "Chicken",
  },
  {
    id: 8,
    title: "Alfredo Pasta",
    slug: "alfredo-pasta",
    authorId: 1,
    category: "Pasta",
  },
  {
    id: 9,
    title: "Peppered Spaghetti",
    slug: "peppered-spaghetti",
    authorId: 3,
    category: "Pasta",
  },
  {
    id: 10,
    title: "Afang Soup",
    slug: "afang-soup",
    authorId: 3,
    category: "Soup",
  },
];

const authorData = [
  { id: 1, name: "Dan Wilder" },
  { id: 2, name: "Oprah Winphrey" },
  { id: 3, name: "Father Abraham" },
];

const newRecipe = {
  id: null,
  title: "",
  authorId: null,
  category: "",
};

// Using CommonJS style export so we can consume via Node (without using Babel-node)
module.exports = {
  newRecipe,
  recipeData,
  authorData,
};
