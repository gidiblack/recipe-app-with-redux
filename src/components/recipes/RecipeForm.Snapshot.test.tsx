import React from "react";
import RecipeForm from "./RecipeForm";
import renderer from "react-test-renderer";
import { recipeData, authorData } from "./../../../utils/mockData";

it("sets submit button label to 'Saving...' when saving is true", () => {
  // call renderer to render the component. renderer returns a tree which is an object that represents the ouput of our component
  // jest.fn() adds an empty function to the test
  const tree = renderer.create(
    <RecipeForm
      recipe={recipeData[0]}
      authors={authorData}
      onSave={jest.fn()}
      onChange={jest.fn()}
      saving
      errors
    />
  );

  expect(tree).toMatchSnapshot();
});

it("sets submit button label to 'Save' when saving is false", () => {
  // call renderer to render the component. renderer returns a tree which is an object that represents the ouput of our component
  // jest.fn() adds an empty function to the test
  const tree = renderer.create(
    <RecipeForm
      recipe={recipeData[1]}
      authors={authorData}
      onSave={jest.fn()}
      onChange={jest.fn()}
      saving={false}
      errors
    />
  );

  expect(tree).toMatchSnapshot();
});
