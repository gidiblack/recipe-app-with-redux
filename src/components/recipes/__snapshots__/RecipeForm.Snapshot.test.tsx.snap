// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sets submit button label to 'Save' when saving is false 1`] = `
<form
  onSubmit={[MockFunction]}
>
  <h2>
    Edit
     Recipe
  </h2>
  <div
    className="form-group"
  >
    <label
      htmlFor="title"
    >
      Title
    </label>
    <div
      className="field"
    >
      <input
        className="form-control my-2"
        name="title"
        onChange={[MockFunction]}
        placeholder="Enter Recipe Title"
        type="text"
        value="Egusi Soup"
      />
    </div>
  </div>
  <div>
    <label
      htmlFor="authorId"
    >
      Author
    </label>
    <div
      className="field"
    >
      <select
        className="form-control my-2"
        name="authorId"
        onChange={[MockFunction]}
        value={3}
      >
        <option
          value=""
        >
          Select Author
        </option>
        <option
          value={1}
        >
          Dan Wilder
        </option>
        <option
          value={2}
        >
          Oprah Winphrey
        </option>
        <option
          value={3}
        >
          Father Abraham
        </option>
      </select>
    </div>
  </div>
  <div
    className="form-group"
  >
    <label
      htmlFor="category"
    >
      Category
    </label>
    <div
      className="field"
    >
      <input
        className="form-control my-2"
        name="category"
        onChange={[MockFunction]}
        placeholder="Enter Category"
        type="text"
        value="Soup"
      />
    </div>
  </div>
  <button
    className="btn btn-primary mt-3"
    disabled={false}
    type="submit"
  >
    Save
  </button>
</form>
`;

exports[`sets submit button label to 'Saving...' when saving is true 1`] = `
<form
  onSubmit={[MockFunction]}
>
  <h2>
    Edit
     Recipe
  </h2>
  <div
    className="form-group"
  >
    <label
      htmlFor="title"
    >
      Title
    </label>
    <div
      className="field"
    >
      <input
        className="form-control my-2"
        name="title"
        onChange={[MockFunction]}
        placeholder="Enter Recipe Title"
        type="text"
        value="Grilled fish sticks"
      />
    </div>
  </div>
  <div>
    <label
      htmlFor="authorId"
    >
      Author
    </label>
    <div
      className="field"
    >
      <select
        className="form-control my-2"
        name="authorId"
        onChange={[MockFunction]}
        value={1}
      >
        <option
          value=""
        >
          Select Author
        </option>
        <option
          value={1}
        >
          Dan Wilder
        </option>
        <option
          value={2}
        >
          Oprah Winphrey
        </option>
        <option
          value={3}
        >
          Father Abraham
        </option>
      </select>
    </div>
  </div>
  <div
    className="form-group"
  >
    <label
      htmlFor="category"
    >
      Category
    </label>
    <div
      className="field"
    >
      <input
        className="form-control my-2"
        name="category"
        onChange={[MockFunction]}
        placeholder="Enter Category"
        type="text"
        value="Fish"
      />
    </div>
  </div>
  <button
    className="btn btn-primary mt-3"
    disabled={true}
    type="submit"
  >
    Saving...
  </button>
</form>
`;
